setSelectionRange will reset the m_latestFocusTrigger variable, causing the subsequently focused element to match the :focus-visible selector

ntdiary opened the pull request:
#### 3d85c92f718f3ce85c7e71c800810dfde1669403
<pre>
setSelectionRange will reset the m_latestFocusTrigger variable, causing the subsequently focused element to match the :focus-visible selector
<a href="https://bugs.webkit.org/show_bug.cgi?id=256929">https://bugs.webkit.org/show_bug.cgi?id=256929</a>
&lt;<a href="https://rdar.apple.com/problem/109803748">rdar://problem/109803748</a>&gt;

Reviewed by Ryosuke Niwa.

This patch adds ForBindings to differentiate when setSelectionRange is called from Javascript.
In this way, if it is called from Javascript and focus occurs,
its behavior can be consistent with Element::focusForBindings.

* LayoutTests/fast/selectors/focus-visible-setSelectionRange-expected.txt: Added.
* LayoutTests/fast/selectors/focus-visible-setSelectionRange.html: Added.
* Source/WebCore/editing/FrameSelection.cpp:
(WebCore::FrameSelection::moveWithoutValidationTo):
(WebCore::FrameSelection::setSelectionWithoutUpdatingAppearance):
(WebCore::FrameSelection::setFocusedElementIfNeeded):
* Source/WebCore/editing/FrameSelection.h:
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::setSelectionRangeForBindings):
* Source/WebCore/html/HTMLTextAreaElement.cpp:
(WebCore::HTMLTextAreaElement::setSelectionRangeForBindings):
* Source/WebCore/html/HTMLTextAreaElement.h:
* Source/WebCore/html/HTMLTextAreaElement.idl:
* Source/WebCore/html/HTMLTextFormControlElement.cpp:
(WebCore::HTMLTextFormControlElement::setSelectionRange):
* Source/WebCore/html/HTMLTextFormControlElement.h:

Canonical link: <a href="https://commits.webkit.org/270043@main">https://commits.webkit.org/270043@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/94404dbd761c3a62eba367d8725d00c98310de87

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/23795 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/1908 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/24892 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/25943 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/21943 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/24070 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/3500 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/24295 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/22470 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/24038 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/1445 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/20570 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/26537 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/1218 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/21481 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/27736 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/21691 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/21757 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/25517 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/1177 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/18872 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/1180 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/21253 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5839 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/1589 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/1496 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/editing/FrameSelection.cpp:
```diff
@@ -2353,7 +2353,7 @@ void FrameSelection::setFocusedElementIfNeeded()
             // work in the long term, but this is the safest fix at this time.
             if (target->isMouseFocusable() && !isFrameElement(target)) {
-                CheckedRef(m_document->page()->focusController())->setFocusedElement(target, *m_document->frame());
+                CheckedRef(m_document->page()->focusController())->setFocusedElement(target, *m_document->frame(), { { }, { }, { }, FocusTrigger::Bindings, { } });
```
This is not bindings. Why are we using FocusTrigger::Bindings?

ntdiary replied:
https://github.com/WebKit/WebKit/blob/5512c619b8aa9d8129d15c540158fb6bf32bbee7/Source/WebCore/dom/Document.cpp#L4988-L4989
Now, only `FocusTrigger::Bindings` won't reset the `m_latestFocusTrigger` variable.If we feel it does not conform to the semantics, how about trying to add a new enum value?

rniwa replied:
We should probably do that. But more importantly, why is not resetting m_latestFocusTrigger right for all calls of this function? This function is called whenever selection is changed. If the problem here lies when this function is called by setSelectionRange then we need a more targeted code change instead.

ntdiary replied:
Sorry for the delay, I'm still delving deeper into the relevant code and change history, hope to have a good understanding by this weekend. : )

rniwa commented on Source/WebCore/html/HTMLTextFormControlElement.h:
```diff
@@ -144,6 +144,7 @@ class HTMLTextFormControlElement : public HTMLFormControlElement {
     void internalSetMaxLength(int maxLength) { m_maxLength = maxLength; }
     void internalSetMinLength(int minLength) { m_minLength = minLength; }
 
+    bool isForBindings { false };
```
Please don't add a new public boolean member variable like this.
Instead, we should be passing this argument across function calls.

ntdiary replied:
> Please don't add a new public boolean member variable like this. Instead, we should be passing this argument across function calls.

Thank you very much for your advice! I didn't know about this before and will try to remove it later.

ntdiary replied:
Hi, @rniwa, I mistakenly thought we would not want to change the function signature, so I added that `protected` member variable. This is my first WebKit PR, not too familiar with the coding conventions yet 😅. 

> we should be passing this argument across function calls.

Does this mean I can append an argument to these functions?
```c++
// HTMLTextFormControlElement.h
void setSelectionRange(unsigned start, unsigned end, const String& direction, const AXTextStateChangeIntent& = AXTextStateChangeIntent());
WEBCORE_EXPORT bool setSelectionRange(unsigned start, unsigned end, TextFieldSelectionDirection = SelectionHasNoDirection, SelectionRevealMode = SelectionRevealMode::DoNotReveal, const AXTextStateChangeIntent& = AXTextStateChangeIntent());
```
Or we just should avoid public member variable (but allow private one), then I can add a new `protected` function and have `input` and `textarea` call it?
```c++
// HTMLTextFormControlElement.h
void setSelectionRange(unsigned start, unsigned end, const String& direction, bool isForBindings = false);
```

rniwa replied:
Adding a new argument to the function seems fine. Don't add a new state to HTMLTextFormControlElement, let alone a public state variable.

rniwa reviewed (changes requested):
r- because we don't want to be adding a new boolean member to `HTMLTextFormControlElement` like this.

rniwa commented on Source/WebCore/editing/FrameSelection.h:
```diff
@@ -359,6 +359,8 @@ class FrameSelection final : private CaretBase, public CaretAnimationClient {
     bool m_pendingSelectionUpdate : 1;
     bool m_alwaysAlignCursorOnScrollWhenRevealingSelection : 1;
 
+    bool m_isForBindings { false };
```
Please don't add a new state like this to FrameSelection.

rniwa reviewed (changes requested):
r- because of a new state being added to FrameSelection.

rniwa commented on LayoutTests/fast/selectors/focus-visible-setSelectionRange-expected.txt:
```diff
@@ -0,0 +1,6 @@
+Click me
```
Why there is no output for this test??

ntdiary replied:
Hi, could you explain what "output" means? 🤔

rniwa replied:
Hm... there is output now. I guess it was an issue with my browser.

rniwa commented on Source/WebCore/editing/FrameSelection.cpp:
```diff
@@ -243,7 +243,7 @@ void FrameSelection::moveTo(const Position& base, const Position& extent, Affini
 }
 
-void FrameSelection::moveWithoutValidationTo(const Position& base, const Position& extent, bool selectionHasDirection, bool shouldSetFocus, SelectionRevealMode revealMode, const AXTextStateChangeIntent& intent)
+void FrameSelection::moveWithoutValidationTo(const Position& base, const Position& extent, bool selectionHasDirection, bool shouldSetFocus, SelectionRevealMode revealMode, const AXTextStateChangeIntent& intent, bool isForBindings)
```
Please add an enum class like this:
```
enum class ForBindings : bool { No, Yes };
```

rniwa commented on Source/WebCore/editing/FrameSelection.h:
```diff
@@ -144,10 +144,10 @@ class FrameSelection final : private CaretBase, public CaretAnimationClient {
     void moveTo(const Position&, Affinity, UserTriggered = UserTriggered::No);
     void moveTo(const Position&, const Position&, Affinity, UserTriggered = UserTriggered::No);
-    void moveWithoutValidationTo(const Position&, const Position&, bool selectionHasDirection, bool shouldSetFocus, SelectionRevealMode, const AXTextStateChangeIntent& = AXTextStateChangeIntent());
+    void moveWithoutValidationTo(const Position&, const Position&, bool selectionHasDirection, bool shouldSetFocus, SelectionRevealMode, const AXTextStateChangeIntent& = AXTextStateChangeIntent(), bool isForBindings = false);
```
Other than adding a new boolean, we should probably add a new value of `SetSelectionOption` instead.
e.g. `SetSelectionOption::IsScriptTriggered` or `SetSelectionOption::ForBindings`

ntdiary replied:
To distinguish it from the existing `SetSelectionOption::IsUserTriggered`, I chose `SetSelectionOption::ForBindings`.

ntdiary replied:
As `moveWithoutValidationTo` is only called in `HTMLTextFormControlElement.cpp`, and `shouldSetFocus` and `SelectionRevealMode` are both used to set `SetSelectionOption`, I move the logic for these two parameters into `HTMLTextFormControlElement.cpp`. If you think we should not do this, please feel free to let me know. : )

rniwa commented on LayoutTests/fast/selectors/focus-visible-setSelectionRange.html:
```diff
@@ -0,0 +1,41 @@
+        }, 'Script focus after setSelectionRange after mouse focus does not match :focus-visible');
+
+        if (window.testRunner) {
+            UIHelper.activateElement(initial);
```
No curly braces around a single line statement.

ntdiary committed:
setSelectionRange will reset the m_latestFocusTrigger variable, causing the subsequently focused element to match the :focus-visible selector
https://bugs.webkit.org/show_bug.cgi?id=256929
<rdar://problem/109803748>

Reviewed by Ryosuke Niwa.

This patch adds ForBindings to differentiate when setSelectionRange is called from Javascript.
In this way, if it is called from Javascript and focus occurs,
its behavior can be consistent with Element::focusForBindings.

* LayoutTests/fast/selectors/focus-visible-setSelectionRange-expected.txt: Added.
* LayoutTests/fast/selectors/focus-visible-setSelectionRange.html: Added.
* Source/WebCore/editing/FrameSelection.cpp:
(WebCore::FrameSelection::moveWithoutValidationTo):
(WebCore::FrameSelection::setSelectionWithoutUpdatingAppearance):
(WebCore::FrameSelection::setFocusedElementIfNeeded):
* Source/WebCore/editing/FrameSelection.h:
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::setSelectionRangeForBindings):
* Source/WebCore/html/HTMLTextAreaElement.cpp:
(WebCore::HTMLTextAreaElement::setSelectionRangeForBindings):
* Source/WebCore/html/HTMLTextAreaElement.h:
* Source/WebCore/html/HTMLTextAreaElement.idl:
* Source/WebCore/html/HTMLTextFormControlElement.cpp:
(WebCore::HTMLTextFormControlElement::setSelectionRange):
* Source/WebCore/html/HTMLTextFormControlElement.h:

Canonical link: https://commits.webkit.org/270043@main
